tests/add_tasks_test.go

```diff
@@ -0,0 +1,222 @@
+// that workflow task is retried.
+func (e *noopExecutable) Execute() error {
+	task := e.GetTask()
+	if e.shouldExecute(task) {
+		return e.Executable.Execute()
+	}
+	// If we don't execute the task, just record it.
+	e.s.skippedTasks <- task
+	return nil
+}
+
+// shouldExecute returns true if the task is not a workflow task, or if the workflow task is not from this test suite
+// (e.g. from the history scanner), or if we've turned off skipping (which we do when we re-add the task).
+func (e *noopExecutable) shouldExecute(task tasks.Task) bool {
+	return task.GetWorkflowID() != e.s.workflowID ||
+		task.GetType() != enumspb.TASK_TYPE_TRANSFER_WORKFLOW_TASK ||
+		!e.s.shouldSkip
+}
+
+// SetupSuite creates the test cluster and registers the executableWrapper with the history service.
+func (s *addTasksSuite) SetupSuite() {
+	// We do this here and in SetupTest because we need assertions in the SetupSuite method as well as the individual
+	// tests, but this is called before SetupTest, and the s.T() value will change when SetupTest is called.
+	s.Assertions = require.New(s.T())
+	// Set up the test cluster and register our executable wrapper.
+	s.setupSuite("testdata/cluster.yaml",
+		WithFxOptionsForService(
+			primitives.HistoryService,
+			fx.Decorate(func(w queues.ExecutableWrapper) queues.ExecutableWrapper {
+				return &executableWrapper{w: w, s: s}
+			}),
+		),
+	)
+	// Get an SDK client so that we can call ExecuteWorkflow.
+	s.sdkClient = s.newSDKClient()
+}
+
+func (s *addTasksSuite) TearDownSuite() {
+	s.sdkClient.Close()
+	s.tearDownSuite()
+}
+
+func (s *addTasksSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+}
+
+func TestAddTasksSuite(t *testing.T) {
+	suite.Run(t, new(addTasksSuite))
+}
+
+func (s *addTasksSuite) TestAddTasks_Ok() {
+	// Register a workflow which records when it is executed on a channel.
+	taskQueue := s.randomizeStr("add-tasks-test-queue")
+	w := worker.New(s.sdkClient, taskQueue, worker.Options{DeadlockDetectionTimeout: 0})
+	executed := make(chan struct{})
+	myWorkflow := func(ctx workflow.Context) error {
+		executed <- struct{}{}
+		return nil
+	}
+	s.NoError(w.Start())
+	defer w.Stop()
+	w.RegisterWorkflow(myWorkflow)
+
+	// Execute that workflow
+	// We need to track the workflow ID so that we can filter out tasks from this test suite
+	s.workflowID = uuid.New()
+	s.shouldSkip = true
+	s.skippedTasks = make(chan tasks.Task)
+	ctx := context.Background()
+	timeout := 5 * debug.TimeoutMultiplier * time.Second
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	workflowID := s.workflowID
+	run, err := s.sdkClient.ExecuteWorkflow(ctx, sdkclient.StartWorkflowOptions{
+		ID:        workflowID,
+		TaskQueue: taskQueue,
+	}, myWorkflow)
+	s.NoError(err)
+
+	// Verify that the workflow did not complete (because we skipped its workflow task)
+	select {
+	case <-executed:
+		s.FailNow("workflow was executed even though the task was failed")
+	default:
+	}
```

Comment:
this check can still pass even when workflow task is processed? I mean race condition between this check and the transferQ processing.

The test will still fail on L177, which makes ^ check seems unnecessary.

Reply:
Added a counterfactual test which does not call AddTasks and verified that it fails.